micro_sam/training/training.py

```diff
@@ -0,0 +1,117 @@
+        save_root: Optional root directory for saving the checkpoints and logs.
+            If not given the current working directory is used.
+    """
+    # TODO check the data loaders!
+
+    device = get_device(device)
+
+    # Get the trainable segment anything model.
+    model, state = get_trainable_sam_model(
+        model_type=model_type, device=device, freeze=freeze,
+        checkpoint_path=checkpoint_path, return_state=True,
+    )
+
+    # This class creates all the training data for a batch (inputs, prompts and labels).
+    convert_inputs = ConvertToSamInputs(transform=model.transform, box_distortion_factor=0.025)
+
+    # Create the UNETR decoder (if train with it) and the optimizer.
+    if with_segmentation_decoder:
+
+        # Get the UNETR.
+        unetr = get_unetr(
+            image_encoder=model.sam.image_encoder,
+            decoder_state=state.get("decoder_state", None),
+            device=device,
+        )
+
+        # Get the parameters for SAM and the decoder from UNETR.
+        joint_model_params = [params for params in model.parameters()]  # sam parameters
+        for param_name, params in unetr.named_parameters():  # unetr's decoder parameters
+            if not param_name.startswith("encoder"):
+                joint_model_params.append(params)
+
+        optimizer = torch.optim.Adam(joint_model_params, lr=lr)
+
+    else:
+        optimizer = torch.optim.Adam(model.parameters(), lr=lr)
+
+    if scheduler is None:
+        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
+            optimizer, mode="min", factor=0.9, patience=3, verbose=True
+        )
+
+    # The trainer which performs training and validation.
+    if with_segmentation_decoder:
+        instance_seg_loss = DiceBasedDistanceLoss(mask_distances_in_bg=True)
+        trainer = joint_trainers.JointSamTrainer(
+            name=name, train_loader=train_loader, val_loader=val_loader, model=model,
+            optimizer=optimizer, device=device, lr_scheduler=scheduler, logger=joint_trainers.JointSamLogger,
+            log_image_interval=100, mixed_precision=True, convert_inputs=convert_inputs,
+            n_objects_per_batch=n_objects_per_batch, n_sub_iteration=n_sub_iteration, compile_model=False, unetr=unetr,
+            instance_loss=instance_seg_loss, instance_metric=instance_seg_loss,
+            early_stopping=early_stopping, mask_prob=0.5, save_root=save_root,
+        )
+    else:
+        trainer = trainers.SamTrainer(
+            name=name, train_loader=train_loader, val_loader=val_loader, model=model,
+            optimizer=optimizer, device=device, lr_scheduler=scheduler, logger=trainers.SamLogger,
+            log_image_interval=100, mixed_precision=True, convert_inputs=convert_inputs,
+            n_objects_per_batch=n_objects_per_batch, n_sub_iteration=n_sub_iteration, compile_model=False,
+            early_stopping=early_stopping, mask_prob=0.5, save_root=save_root,
```

Comment:
Would it also make sense to expose the mask probability ratio? (we don't really need it as much anywhere, but just in case)